instrumentation/net/http/otelhttp/internal/body_wrapper.go

```diff
@@ -0,0 +1,53 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal // import "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/internal"
+
+import (
+	"io"
+	"sync/atomic"
+)
+
+var _ io.ReadCloser = &BodyWrapper{}
+
+// BodyWrapper wraps a http.Request.Body (an io.ReadCloser) to track the number
+// of bytes read and the last error.
+type BodyWrapper struct {
+	io.ReadCloser
+	OnRead func(n int64) // must not be nil
+
+	read atomic.Int64
+	err  atomic.Value
+}
+
+// Read reads the data from the io.ReadCloser, and stores the number of bytes
+// read and the error.
+func (w *BodyWrapper) Read(b []byte) (int, error) {
+	n, err := w.ReadCloser.Read(b)
+	n1 := int64(n)
+	w.read.Add(n1)
+	if err != nil {
+		w.err.Store(err)
```

Comment:
["All calls to Store for a given Value must use values of the same concrete type."](https://pkg.go.dev/sync/atomic#Value.Store) This will panic when different error implementations are stored.

Reply:
Nice catch, thank you.